### PR TITLE
Nfs feature vfd swmr

### PR DESCRIPTION
**Pull request by vchoi-hdfgroup**

### CODE REVIEW

File: src/H5Fint.c [hunk 1]
```diff
@@ -1835,25 +1835,26 @@ H5F__check_if_using_file_locks(H5P_genplist_t *fapl, hbool_t *use_file_locking)
...
+    unsigned                page_buf_min_raw_perc  = 0;
+    hbool_t                 set_flag               = FALSE; /* Set the status_flags in the superblock */
+    hbool_t                 clear                  = FALSE; /* Clear the status_flags */
+    hbool_t                 evict_on_close;                 /* Evict on close value from plist */
+    hbool_t                 use_file_locking    = TRUE;     /* Using file locks? */
+    hbool_t                 ci_load             = FALSE;    /* Whether MDC ci load requested */
+    hbool_t                 ci_write            = FALSE;    /* Whether MDC CI write requested */
+    hbool_t                 file_create         = FALSE;    /* Creating a new file or not */
+    H5F_vfd_swmr_config_t * vfd_swmr_config_ptr = NULL;     /* Points to VFD SMWR config info */
+    H5F_generate_md_ck_cb_t cb_info             = {NULL};
```
Comment (kyang2014) [inline src/H5Fint.c]: May add a comment on what this callback is for like the other variables?

File: src/H5Fint.c [hunk 2]
```diff
@@ -1835,25 +1835,26 @@ H5F__check_if_using_file_locks(H5P_genplist_t *fapl, hbool_t *use_file_locking)
...
+    H5F_close_degree_t      fc_degree;     /*file close degree             */
+    size_t                  page_buf_size;
+    unsigned                page_buf_min_meta_perc = 0;
+    unsigned                page_buf_min_raw_perc  = 0;
+    hbool_t                 set_flag               = FALSE; /* Set the status_flags in the superblock */
+    hbool_t                 clear                  = FALSE; /* Clear the status_flags */
+    hbool_t                 evict_on_close;                 /* Evict on close value from plist */
+    hbool_t                 use_file_locking    = TRUE;     /* Using file locks? */
+    hbool_t                 ci_load             = FALSE;    /* Whether MDC ci load requested */
+    hbool_t                 ci_write            = FALSE;    /* Whether MDC CI write requested */
```
Comment (kyang2014) [inline src/H5Fint.c]: Minor: CI in the comment may ci like the last line

File: src/H5Fprivate.h [hunk 3]
```diff
@@ -784,6 +789,72 @@ uint64_decode(uint8_t **pp)
...
+
+/* Size of updater file header */
+#define H5F_UD_HEADER_SIZE                                                                                   \
+    (H5_SIZEOF_MAGIC     /* Signature */                                                                     \
+     + 2                 /* Version number */                                                                \
+     + 2                 /* Flags */                                                                         \
+     + 4                 /* Page size */                                                                     \
+     + 8                 /* Sequence number */                                                               \
+     + 8                 /* Tick number */                                                                   \
+     + 8                 /* Chnage list offset */                                                            \
```
Comment (kyang2014) [inline src/H5Fprivate.h]: Minor: typo - Chnage should be Change

File: src/H5Fprivate.h [hunk 4]
```diff
@@ -854,6 +925,180 @@ typedef enum H5F_prefix_open_t {
...
+ * VFD SWMR
+ */
+
+/*----------------------------------------------------------------------------
+ *
+ *  struct H5F_vfd_swmr_updater_cl_entry_t
+ *
+ *  An array of instances of H5F_vfd_swmr_updater_cl_entry_t of length equal to
+ *  the number of metadata pages and multi-page metadata entries modified in
+ *  the past tick is used ot aseemble the assoicated data in preparation for
```
Comment (kyang2014) [inline src/H5Fprivate.h]: Minor typo: "used ot aseemble the assoicated" should be "used to assemble the associated"

File: src/H5Fprivate.h [hunk 5]
```diff
@@ -854,6 +925,180 @@ typedef enum H5F_prefix_open_t {
...
+ *           metadata file index in the updater file.
+ *
+ *  md_file_index_len: This field contains the size of the metadata file index in
+ *           bytes.
+ *
+ *  num_change_list_entries: This field contains the number of entries in the
+ *           array of H5F_vfd_swmr_updater_cl_ entry_t whose base address
+ *           is stored in the change_list field below.  This value is also the
+ *           number of metadata pages and multi-page metadata entries that have
+ *           been modified in the past tick
```
Comment (kyang2014) [inline src/H5Fprivate.h]: Minor add a '.' at the end of "past tick"

File: src/H5Fvfd_swmr.c [hunk 6]
```diff
@@ -2010,3 +2059,388 @@ H5F_post_vfd_swmr_log_entry(H5F_t *f, int entry_type_code, char *log_info)
...
+    else {
+        /* Update 2 updater file header fields: change_list_len, change_list_offset */
+        updater.change_list_len    = H5F_UD_CL_SIZE(updater.num_change_list_entries);
+        updater.change_list_offset = updater.header_image_len;
+    }
+
+    /* Create the updater file with a temporary file name */
+    HDsprintf(namebuf, "%s.ud_tmp", shared->vfd_swmr_config.updater_file_path);
+
+    if ((ud_file = H5FD_open(namebuf, H5F_ACC_TRUNC | H5F_ACC_RDWR | H5F_ACC_CREAT, H5P_FILE_ACCESS_DEFAULT,
```
Comment (kyang2014) [inline src/H5Fvfd_swmr.c:2307]: I must miss something here.  I think the updater file is not an HDF5 file. Why calling H5FD_open() here? will the H5FD_open() eventually call the VFD SMWR open call?

File: test/vfd_swmr.c [hunk 7]
```diff
@@ -3424,6 +3686,701 @@ test_enable_disable_eot(void)
...
+ *              1 if test fails
+ *
+ * Programmer:  Vailin Choi; October 2021
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+verify_updater_flags(char *ud_name, uint16_t expected_flags)
+{
+    FILE *   ud_fp; /* Updater file pointer */
```
Comment (kyang2014) [inline test/vfd_swmr.c]: Minor: should we initialize the three local variables like the following routine?

File: test/vfd_swmr.c [hunk 8]
```diff
@@ -3424,6 +3686,701 @@ test_enable_disable_eot(void)
...
+
+    TESTING("VFD SWMR updater file flags for multiple opens of the same file");
+
+    /* Should succeed without VFD SWMR configured */
+    if ((fid = H5Fcreate(FILENAME, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT)) < 0)
+        TEST_ERROR;
+
+    /* Close the file  */
+    if (H5Fclose(fid) < 0)
+        FAIL_STACK_ERROR;
```
Comment (kyang2014) [inline test/vfd_swmr.c:3911]: Minor: I see "FAIL_STACk_ERROR" a few lines below, should the ";" be removed? The same as "TEST_ERROR;" a few lines above

File: test/vfd_swmr.c [hunk 9]
```diff
@@ -3424,6 +3686,701 @@ test_enable_disable_eot(void)
...
+ * Return:      void
+ *
+ * Programmer:  Vailin Choi; October 2021
+ *
+ *-------------------------------------------------------------------------
+ */
+static void
+clean_chk_ud_files(char *md_file_path, char *updater_file_path)
+{
+    char     chk_name[1024 + 4];
```
Comment (kyang2014) [inline test/vfd_swmr.c]: Can the [1024 +4 ] be replaced by a constant +4  or it doesn't matter?

File: test/vfd_swmr.c [hunk 10]
```diff
@@ -3424,6 +3686,701 @@ test_enable_disable_eot(void)
...
+    /* Remove the checksum file if exists.
+       If not, the callback will just continue appending
+       checksums to the existing file */
+    if (HDaccess(chk_name, F_OK) == 0) {
+        HDremove(chk_name);
+    }
+
+    /* Remove all the updater files if exist: <updater_file_path>.<i> */
+    for (i = 0;; i++) {
+        sprintf(ud_name, "%s.%lu", updater_file_path, i);
```
Comment (kyang2014) [inline test/vfd_swmr.c]: May still use HDsprintf.

File: test/vfd_swmr.c [hunk 11]
```diff
@@ -3424,6 +3686,701 @@ test_enable_disable_eot(void)
...
+ *              1 if test fails
+ *
+ * Programmer:  Vailin Choi; October 2021
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+verify_ud_chk(char *md_file_path, char *ud_file_path)
+{
+    char     chk_name[1024 + 4];      /* Checksum file name */
```
Comment (kyang2014) [inline test/vfd_swmr.c]: Minor: Should we use a constant to replace 1024 here? Also should the local variables be initialized? Should the ";" in 'FAIL_STACK_ERROR; ' and 'TEST_ERROR;' be removed or make them consistent?

File: test/vfd_swmr.c [hunk 12]
```diff
@@ -3424,6 +3686,701 @@ test_enable_disable_eot(void)
...
+
+} /* verify_ud_chk() */
+
+/*-------------------------------------------------------------------------
+ * Function:    md_ck_cb()
+ *
+ * Purpose:     This is the callback function used by
+ *              test_updater_generate_md_checksums() when the
+ *              H5F_ACS_GENERATE_MD_CK_CB_NAME property is set in fapl.
+ *                  --Opens and read the metadata file into a buffer.
```
Comment (kyang2014) [inline test/vfd_swmr.c]: Typo: Open instead of Opens

File: test/vfd_swmr.c [hunk 13]
```diff
@@ -3424,6 +3686,701 @@ test_enable_disable_eot(void)
...
+    if ((fcpl = vfd_swmr_create_fcpl(H5F_FSPACE_STRATEGY_PAGE, 4096)) < 0) {
+        HDprintf("vfd_swmr_create_fcpl() failed");
+        FAIL_STACK_ERROR
+    }
+
+    /* Set up callback to generate checksums for updater's metadata files */
+    cb_info.func = md_ck_cb;
+
+    /* Activate private property to generate checksums for updater's metadata file */
+    H5Pset(fapl, H5F_ACS_GENERATE_MD_CK_CB_NAME, &cb_info);
```
Comment (kyang2014) [inline test/vfd_swmr.c]: Should we check the return value?

File: test/vfd_swmr.c [hunk 14]
```diff
@@ -3424,6 +3686,701 @@ test_enable_disable_eot(void)
...
+
+    /* Use file creation or file open for testing */
+    if (file_create) {
+        if ((fid = H5Fcreate(FILENAME4, H5F_ACC_TRUNC, fcpl, fapl)) < 0)
+            TEST_ERROR;
+    }
+    else {
+        fid = H5Fcreate(FILENAME4, H5F_ACC_TRUNC, fcpl, H5P_DEFAULT);
```
Comment (kyang2014) [inline test/vfd_swmr.c]: Should we check the returned values of APIs in this block?

**Review by kyang2014 — APPROVED** [review]
My comments are mostly trivial. After Vailin check them and make change as appropriate. I don't need to review them again.

File: src/H5Fint.c (hunk 1 above)
Comment (vchoi-hdfgroup) [reply src/H5Fint.c]: done

File: src/H5Fint.c (hunk 2 above)
Comment (vchoi-hdfgroup) [reply src/H5Fint.c]: done

File: src/H5Fprivate.h (hunk 3 above)
Comment (vchoi-hdfgroup) [reply src/H5Fprivate.h]: done

File: src/H5Fprivate.h (hunk 4 above)
Comment (vchoi-hdfgroup) [reply src/H5Fprivate.h]: done

File: src/H5Fprivate.h (hunk 5 above)
Comment (vchoi-hdfgroup) [reply src/H5Fprivate.h]: done

File: test/vfd_swmr.c (hunk 7 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr.c]: done

File: test/vfd_swmr.c (hunk 9 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr.c]: done

File: test/vfd_swmr.c (hunk 10 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr.c]: done

File: test/vfd_swmr.c (hunk 12 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr.c]: done

File: test/vfd_swmr.c (hunk 13 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr.c]: done

File: test/vfd_swmr.c (hunk 14 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr.c]: done

File: src/H5Fvfd_swmr.c (hunk 6 above)
Comment (vchoi-hdfgroup) [reply src/H5Fvfd_swmr.c:2307]: According to page 64 in the RFC, item #4 and #5 for the basic sequence of operations, "use the VFD interface for the updater file."

File: test/vfd_swmr.c (hunk 11 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr.c]: Fix the constant for chk_name[].
As for the ";" after FAIL_STACK_ERROR and TEST_ERROR, I will add ";" to be consistent with most of those calls in this file.  I checked the files in the test directory, some of those calls have ";" and some without.

File: test/vfd_swmr.c (hunk 8 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr.c:3911]: I checked the files in the test directory, some of those calls have ";" and some without. I will add ";" to be consistent with most of those calls in this file test/vfd_swmr.c.

**Review by raylu-hdf — APPROVED** [review]
I tried this branch.  It worked with some basic tests.

**Pull request merged**